api/provider/aws/environment_scaler.go

```diff
@@ -1,14 +1,758 @@
+	if err != nil {
+		return err
+	}
+
+	resourceConsumers = append(resourceConsumers, jobTaskConsumers...)
+
+	log.Printf("[DEBUG] [EnvironmentScaler] resourceConsumers for env '%s': %#v", environmentID, resourceConsumers)
+
+	var errs []error
+
+	// calculate for scaling up
+	resourceProviders, scaleUpErrs := e.calculateScaleUp(clusterName, resourceProviders, resourceConsumers)
+
+	errs = append(errs, scaleUpErrs...)
+
+	// calculate for scaling down
+	unusedProviders := e.calculateScaleDown(clusterName, resourceProviders)
+
+	// do the scaling
+	desiredScale := len(resourceProviders) - len(unusedProviders)
+
+	if err := e.scaleTo(clusterName, desiredScale, unusedProviders); err != nil {
+		errs = append(errs, err)
+	}
+
+	return errors.MultiError(errs)
+}
+
+func (e *EnvironmentScaler) calculateNewProvider(clusterName string) (*ResourceProvider, error) {
+	env, err := e.EnvironmentProvider.Read(clusterName)
+	if err != nil {
+		return nil, err
+	}
+
+	resource := &ResourceProvider{}
+	resource.AvailableMemory = InstanceSizes[env.InstanceSize]
+	resource.ID = "<new instance>"
+	resource.InUse = false
+	resource.UsedPorts = defaultPorts
+
+	return resource, nil
+}
+
+func (e *EnvironmentScaler) calculateScaleDown(clusterName string, resourceProviders []*ResourceProvider) []*ResourceProvider {
+	var unusedProviders []*ResourceProvider
+
+	for _, provider := range resourceProviders {
+		if !provider.InUse {
+			unusedProviders = append(unusedProviders, provider)
+		}
+	}
+
+	return unusedProviders
+}
+
+func (e *EnvironmentScaler) calculateScaleUp(clusterName string, resourceProviders []*ResourceProvider, resourceConsumers []ResourceConsumer) ([]*ResourceProvider, []error) {
+	sorts := []struct {
+		Errs      []error
+		Providers []*ResourceProvider
+		SortBy    string
+	}{
+		{[]error{}, resourceProviders, "cpu"},
+		{[]error{}, resourceProviders, "mem"},
+	}
+
+	for _, s := range sorts {
+		switch s.SortBy {
+		case "cpu":
+			sortProvidersByCPU(s.Providers)
+			sortConsumersByCPU(resourceConsumers)
+
+		case "mem":
+			sortProvidersByMemory(s.Providers)
+			sortConsumersByMemory(resourceConsumers)
+		}
+
+		sortProvidersByUsage(s.Providers)
+
+		for _, consumer := range resourceConsumers {
+			hasRoom := false
+
+			for _, provider := range s.Providers {
+				if provider.HasResourcesFor(consumer) {
+					hasRoom = true
+					provider.SubtractResourcesFor(consumer)
+					break
+				}
+			}
+
+			if !hasRoom {
+				newProvider, err := e.calculateNewProvider(clusterName)
+				if err != nil {
+					s.Errs = append(s.Errs, err)
+					continue
+				}
+
+				if !newProvider.HasResourcesFor(consumer) {
+					text := fmt.Sprintf("Resource '%s' cannot fit into an empty provider!", consumer.ID)
+					text += "\nThe instance size in your environment is too small to run this resource."
+					text += "\n Please increase the instance size for your environment."
+					err := fmt.Errorf(text)
+					s.Errs = append(s.Errs, err)
```

Comment:
This, to me, seems the only plausible use-case for multierrors. Reason being: if you have 1 problematic resource consumer (e.g. someone wants to run a task with a billion units of memory), then we can log it and continue calculating scale for everything else. The problematic consumer will never go away until the user deletes the service/job trying to start it. 

That being said, this code isn't properly setup to handle that type of error; the caller either returns the error that was returned or blindly continues in the face of errors. Both options aren't great. One solution is to create a new type of error, `RecoverableError`, that the caller can check for.

Reply:
Overhauled error handling in this part of the code. ✅